activesupport/lib/active_support/message_verifier.rb

```diff
@@ -64,54 +66,184 @@ def valid_message?(signed_message)
+    # Decodes the signed message using the +MessageVerifier+'s secret. It also decodes
+    # legacy signed messages which are outdated messages generated by +MessageVerifier+
+    # of the previous rails version.
     #
     #   verifier = ActiveSupport::MessageVerifier.new 's3Krit'
-    #   signed_message = verifier.generate 'a private message'
+    #   signed_message = verifier.generate('a private message', for: 'test', expires_in: 1.day)
     #
-    #   verifier.verify(signed_message) # => 'a private message'
+    #   # Within 1 day...
+    #   verifier.verify(signed_message, for: 'test') # => "a private message"
+    #
+    #   legacy_signed_message = "BAhJIhZhIHByaXZhdGUgbWVzc2FnZQY6BkVU--a2a1232deaf49ba164ed0d0b43b3f17cf8026c46"
+    #   verifier.verify(legacy_signed_message) # => "a private message"
     #
     # Raises +InvalidSignature+ if the message was not signed with the same
     # secret or was not Base64-encoded.
     #
     #   other_verifier = ActiveSupport::MessageVerifier.new 'd1ff3r3nt-s3Krit'
-    #   other_verifier.verify(signed_message) # => ActiveSupport::MessageVerifier::InvalidSignature
-    def verify(signed_message)
-      verified(signed_message) || raise(InvalidSignature)
+    #   other_verifier.verify(signed_message) # raises ActiveSupport::MessageVerifier::InvalidSignature
+    #
+    # Raises +InvalidClaims+ if the message doesn't have the same purpose as that
+    # given in the +options+.
+    #
+    #   verifier.verify(signed_message, for: 'something_else') # raises ActiveSupport::Claims::InvalidClaims
+    #
+    # Raises +ExpiredClaims+ if the message is expired.
+    #
+    #   expired_message = verifier.generate('a private message', expires_at: 1.day.ago)
+    #   verifier.verify(signed_message) # raises ActiveSupport::Claims::ExpiredClaims
+    def verify(signed_message, options = {})
+      if value = verified(signed_message, options)
+        value
+      else
+        unless valid_message?(signed_message) && Claims.verify!(get_claims(signed_message), options)
+          raise InvalidSignature
+        end
+      end
     end
 
-    # Generates a signed message for the provided value.
+    # Generates a signed message for the given +value+ and +options+.
     #
     # The message is signed with the +MessageVerifier+'s secret. Without knowing
-    # the secret, the original value cannot be extracted from the message.
+    # the secret, the original value cannot be extracted from the message. A
+    # message is signed as a JSON Web Token (JWT). The final message is
+    # concatenation of base64-encoded JWS header, base64-encoded JWT claims set
+    # and base64-encoded signature separated by a period ('.'). This format is
+    # known as a JSON Web Signature (JWS):
+    # <base64-encoded header>.<base64-encoded claims>.<base64-encoded signature>
+    #
+    # JSON Web Token (JWT) - A string representing a set of claims as a JSON
+    # object that is encoded in a JWS or JWE, enabling the claims to be
```

Comment:
Maybe you can just specify that JWE means "JSON Web Encryption" ; the name seems clear enough to avoid explanations as you did for JWT and JWS.